unreal/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+//! API to process Unreal Engine 4 crashes
+#![warn(missing_docs)]
+
+extern crate byteorder;
+extern crate bytes;
+extern crate compress;
+extern crate failure;
+
+use std::io::{self, Cursor, Read};
+
+use bytes::{Buf, Bytes};
+use compress::zlib;
+use failure::Fail;
+
+struct Header {
+    pub directory_name: String,
+    pub file_name: String,
+    pub uncompressed_size: i32,
+    pub file_count: i32,
+}
+
+/// Meta-data about a file within a UE4 crash file
+#[derive(Debug)]
+pub struct CrashFileMeta {
+    /// The original index within the UE4 crash file
+    pub index: usize,
+    /// File name
+    pub file_name: String,
+    /// Start of the file within crash dumb
+    pub offset: usize,
+    /// Length of bytes from offset
+    pub len: usize,
+}
+
+/// Errors related to parsing an UE4 crash file
+#[derive(Fail, Debug)]
+pub enum Unreal4ParseError {
+    /// Expected UnrealEngine4 crash (zlib compressed)
+    #[fail(display = "unknown bytes format")]
+    UnknownBytesFormat,
+    /// Value out of bounds
+    #[fail(display = "out of bounds")]
+    OutOfBounds,
+    /// Invalid compressed data
+    #[fail(display = "bad compression")]
+    BadCompression(io::Error),
+}
+
+/// Unreal Engine 4 crash file
+#[derive(Debug)]
+pub struct Unreal4Crash {
+    bytes: Bytes,
+    files: Vec<CrashFileMeta>,
+}
+
+impl Unreal4Crash {
+    /// Creates an instance of `Unreal4Crash` from the original, compressed bytes
+    pub fn from_bytes(bytes: &[u8]) -> Result<Unreal4Crash, Unreal4ParseError> {
+        if bytes.is_empty() {
+            return Err(Unreal4ParseError::UnknownBytesFormat);
```

Comment:
Would be great to return a more specific error here (e.g. `::Empty`). This will make it easier for users and us to understand what went wrong.